fix: auto-heal orphaned function_calls to prevent crash loop

donggyun112 opened the pull request:
**Please ensure you have read the [contribution guide](https://github.com/google/adk-python/blob/main/CONTRIBUTING.md) before creating a pull request.**

### Link to Issue or Description of Change

**1. Link to an existing issue (if applicable):**

- Closes: #3971

**2. Or, if no issue exists, describe the change:**

**Problem:**

When execution is interrupted (e.g., server restart, browser refresh, connection loss) after a `function_call` but before the `function_response` is saved, the session becomes permanently unrecoverable. Anthropic and OpenAI APIs require `tool_calls` to be immediately followed by `tool_results`, so subsequent requests fail with 400 BadRequest, creating a crash loop.

**Solution:**

Detect orphaned function_calls (calls without matching responses) during content processing and inject synthetic error responses to gracefully recover the session.

**Why this approach:**

Two approaches were considered:

| Approach | Description | Pros | Cons |
|----------|-------------|------|------|
| **1. Separation of Concerns** | Separate `_find_orphaned_function_calls()` + `_create_synthetic_response_event()` functions, called after `_rearrange_events_for_async_function_responses_in_history()` | Clear responsibility separation, easier to test independently, self-documenting code | Extra O(N) event iteration, duplicates ID mapping logic already in rearrange function |
| **2. Single-Pass Integration** ✅ | Extend `_rearrange_events_for_async_function_responses_in_history()` with `heal_orphaned_calls` param, detect orphaned calls during existing loop | Reuses existing `function_call_id_to_response_events_index` mapping, no duplicate iteration, better performance | Slightly increases function complexity, mixed responsibilities |

**Decision:** Chose **Approach 2** for the following reasons:
1. The existing `_rearrange_events_for_async_function_responses_in_history()` already builds a `function_call_id_to_response_events_index` mapping - reusing it avoids redundant work
2. Orphaned call detection is logically part of the "rearrangement" process (pairing calls with responses)
3. Avoids extra O(N) iteration over events
4. The `heal_orphaned_calls=False` default maintains backward compatibility

### Testing Plan

**Unit Tests:**

- [x] I have added or updated unit tests for my change.
- [x] All unit tests pass locally.

```bash
$ uv run pytest tests/unittests/flows/llm_flows/test_contents_function.py -v
# 12 passed (7 existing + 5 new)
```

New test cases:
- `test_auto_healing_single_orphaned_function_call` - single orphaned call
- `test_auto_healing_multiple_orphaned_function_calls` - multiple orphaned calls in one event
- `test_auto_healing_partial_orphaned_function_calls` - mix of completed and orphaned calls
- `test_auto_healing_no_healing_when_responses_exist` - no false positives
- `test_auto_healing_logs_warning` - warning log verification

**Manual End-to-End (E2E) Tests:**

Reproduced the issue using a test script that sends broken message history (tool_call without tool_result) to Anthropic/OpenAI/Gemini APIs. Before the fix, all non-Gemini models returned 400 BadRequest. After the fix, the synthetic error response allows the session to continue.

### Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/google/adk-python/blob/main/CONTRIBUTING.md) document.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have manually tested my changes end-to-end.
- [x] Any dependent changes have been merged and published in downstream modules.

### Additional context

**Changes:**

| File | Description |
|------|-------------|
| `src/google/adk/flows/llm_flows/contents.py` | Add `_create_synthetic_response_for_orphaned_calls()` helper, extend `_rearrange_events_for_async_function_responses_in_history()` with `heal_orphaned_calls` parameter |
| `tests/unittests/flows/llm_flows/test_contents_function.py` | Add 5 test cases for auto-healing behavior |

**Key implementation details:**

1. **Synthetic response format**: `{'error': 'Tool execution was interrupted.'}` (follows existing error response pattern in codebase)
2. **Warning log**: `Auto-healing orphaned function_call (id=..., name=...)` for debugging/monitoring
3. **Location**: As specified by maintainer, detection occurs around line 445 in `_get_contents()`

**Known Limitations & Future Work:**

1. **Synthetic responses not persisted to session**: Generated at LLM request time only, not saved to session storage. UI/logs/telemetry may still show orphaned calls as "pending". Future consideration: should synthetic events be persisted? This requires policy decision affecting session history integrity, replay scenarios, and multi-client sync.

2. **Repeated warning logs**: `logger.warning()` emitted each time `_get_contents()` processes an orphaned call. If session resumes multiple times before progressing, same warning repeats. Future options: persist synthetic responses, deduplicate by call ID, or demote to `logger.info()` after first occurrence.

These are intentionally left for future PRs to keep this fix focused and minimal.

ryanaiagent commented:
Hi @donggyun112 , Thank you for your contribution! We appreciate you taking the time to submit this pull request. 
Can you please fix the lint errors before we can proceed with the review. You can use autoformat.sh

donggyun112 committed:
fix: auto-heal orphaned function_calls to prevent crash loop

When execution is interrupted (e.g., server restart or connection loss)
after a function_call but before the function_response is saved, the
session becomes unrecoverable because Anthropic/OpenAI require tool_calls
to be immediately followed by tool_results.

This change detects orphaned function_calls and injects synthetic error
responses to gracefully recover the session.

Changes:
- Add _ORPHANED_CALL_ERROR_RESPONSE constant for error responses
- Add _create_synthetic_response_for_orphaned_calls helper function
- Detect orphaned calls in _rearrange_events_for_async_function_responses_in_history
- Use 'user' as author for synthetic function response events
- Add 5 comprehensive test cases for auto-healing behavior

Fixes #3971

donggyun112 committed:
fix: exclude long-running tools from orphaned call detection

Long-running tools (e.g., human-in-the-loop) intentionally don't produce
immediate function_response events. They should not be treated as orphaned.

Changes:
- Check event.long_running_tool_ids before marking a call as orphaned
- Add tests for long-running tool exclusion

donggyun112 committed:
style: fix pyink formatting in test file

donggyun112 commented:
Found a bug in upstream: BigQuery plugin's `_atexit_cleanup` raises `ReferenceError` when weakref proxy object is already garbage collected. Added a fix to handle this exception.

donggyun112 committed:
refactor: address review feedback - simplify response handling and extract test fixture

- Remove redundant single-item special case in function_response_events handling
- Extract common auto-healing test setup into pytest fixture

donggyun112 committed:
fix: address review feedback and fix failing tests

- Remove redundant single-item special case in function_response handling
- Extract common auto-healing test setup into pytest fixture
- Update compaction tests to account for auto-healing synthetic responses

donggyun112 committed:
Merge branch 'main' into fix/orphaned-tool-calls-crash-loop

donggyun112 committed:
Merge branch 'main' into fix/orphaned-tool-calls-crash-loop

rohityan committed:
Merge branch 'main' into fix/orphaned-tool-calls-crash-loop

rohityan commented:
Hi @donggyun112 , closing this since the underlying issue is fixed.
